agent/engine/engine_sudo_linux_integ_test.go

```diff
@@ -127,3 +151,166 @@ func createTestLocalVolumeMountTask() *apitask.Task {
+	time.Sleep(3 * cfg.TaskCleanupWaitDuration)
+
+	for i := 0; i < 60; i++ {
+		_, ok := taskEngine.(*DockerTaskEngine).State().TaskByArn(testTask.Arn)
+		if !ok {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+	// Make sure all the resource is cleaned up
+	_, err = ioutil.ReadDir(filepath.Join(testDataDir, "logrouter", testTask.Arn))
+	assert.Error(t, err)
+}
+
+func createLogDriverAndContainersUsingLogDriverTask(t *testing.T) *apitask.Task {
+	testTask := createTestTask(validTaskArnPrefix + uuid.New())
+	rawHostConfigInputForLogSender := dockercontainer.HostConfig{
+		LogConfig: dockercontainer.LogConfig{
+			Type: logRouterDriverName,
+			Config: map[string]string{
+				"Name":              "cloudwatch",
+				"exclude-pattern":   "exclude",
+				"include-pattern":   "include",
+				"log_group_name":    testLogGroupName,
+				"log_stream_prefix": testLogGroupPrefix,
+				"region":            testECSRegion,
+				"auto_create_group": "true",
+			},
+		},
+	}
+	rawHostConfigForLogSender, err := json.Marshal(&rawHostConfigInputForLogSender)
+	require.NoError(t, err)
+	testTask.Containers = []*apicontainer.Container{
+		{
+			Name:      "logsender",
+			Image:     testLogSenderImage,
+			Essential: true,
+			// TODO: the log router occasionally failed to send logs when it's shut down very quickly after started.
+			// Let the task run for a while with a sleep helps avoid that failure, but still needs to figure out the
+			// root cause.
+			Command: []string{"sh", "-c", "echo exclude; echo include; sleep 10;"},
+			DockerConfig: apicontainer.DockerConfig{
+				HostConfig: func() *string {
+					s := string(rawHostConfigForLogSender)
+					return &s
+				}(),
+			},
+			DependsOn: []apicontainer.DependsOn{
+				{
+					ContainerName: "logrouter",
+					Condition:     "START",
+				},
+			},
+		},
+		{
+			Name:      "logrouter",
+			Image:     testFluentbitImage,
+			Essential: true,
+			LogRouter: &apicontainer.LogRouter{
+				Type:                 logrouter.LogRouterTypeFluentbit,
+				EnableECSLogMetadata: true,
+			},
+			Environment: map[string]string{
+				"AWS_EXECUTION_ENV": "AWS_ECS_EC2",
+				"FLB_LOG_LEVEL":     "debug",
+			},
+			TransitionDependenciesMap: make(map[apicontainerstatus.ContainerStatus]apicontainer.TransitionDependencySet),
+		},
+	}
+	testTask.ResourcesMapUnsafe = make(map[string][]taskresource.TaskResource)
+	return testTask
+}
+
+func waitCloudwatchLogs(client *cloudwatchlogs.CloudWatchLogs, params *cloudwatchlogs.GetLogEventsInput) (*cloudwatchlogs.GetLogEventsOutput, error) {
+	// The test could fail for timing issue, so retry for 30 seconds to make this test more stable
+	for i := 0; i < 30; i++ {
+		resp, err := client.GetLogEvents(params)
+		if err != nil {
+			awsError, ok := err.(awserr.Error)
+			if !ok || awsError.Code() != "ResourceNotFoundException" {
+				return nil, err
+			}
+		} else if len(resp.Events) > 0 {
+			return resp, nil
+		}
+		time.Sleep(time.Second)
```

Comment:
Would this be better to be exponential backoff? Will this meet throttling issue?

Reply:
it's probably better, but i don't think it's needed for now as many of our functional tests are already calling APIs like this and are not throttled